src/node_quic_session.cc

```diff
@@ -0,0 +1,2953 @@
+                        LABEL, strsize(LABEL),
+                        ctx) != 0) {
+    return -1;
+  }
+
+  return ivlen;
+}
+
+inline ssize_t DeriveHeaderProtectionKey(
+    uint8_t* dest,
+    size_t destlen,
+    const uint8_t* secret,
+    size_t secretlen,
+    const CryptoContext &ctx) {
+  static constexpr uint8_t LABEL[] = "quic hp";
+
+  size_t keylen = aead_key_length(ctx);
+  if (keylen > destlen)
+    return -1;
+
+  if(HKDF_Expand_Label(dest, keylen,
+                       secret, secretlen,
+                       LABEL, strsize(LABEL),
+                       ctx) != 0) {
+    return -1;
+  }
+
+  return keylen;
+}
+
+inline ssize_t UpdateTrafficSecret(
+    uint8_t* dest,
+    size_t destlen,
+    const uint8_t* secret,
+    size_t secretlen,
+    const CryptoContext &ctx) {
+
+  static constexpr uint8_t LABEL[] = "traffic upd";
+
+  if (destlen < secretlen)
+    return -1;
+
+  if (HKDF_Expand_Label(dest, secretlen,
+                        secret, secretlen,
+                        LABEL,
+                        strsize(LABEL),
+                        ctx) != 0) {
+    return -1;
+  }
+
+  return secretlen;
+}
+
+inline int Negotiated_PRF(CryptoContext& ctx, SSL* ssl) {
+  switch (SSL_CIPHER_get_id(SSL_get_current_cipher(ssl))) {
+    case 0x03001301u: // TLS_AES_128_GCM_SHA256
+    case 0x03001303u: // TLS_CHACHA20_POLY1305_SHA256
+      ctx.prf = EVP_sha256();
+      return 0;
+    case 0x03001302u: // TLS_AES_256_GCM_SHA384
+      ctx.prf = EVP_sha384();
+      return 0;
+    default:
+      return -1;
+  }
+}
+
+inline int Negotiated_AEAD(CryptoContext& ctx, SSL* ssl) {
+  switch (SSL_CIPHER_get_id(SSL_get_current_cipher(ssl))) {
+    case 0x03001301u: // TLS_AES_128_GCM_SHA256
+      ctx.aead = EVP_aes_128_gcm();
+      ctx.hp = EVP_aes_128_ctr();
+      return 0;
+    case 0x03001302u: // TLS_AES_256_GCM_SHA384
+      ctx.aead = EVP_aes_256_gcm();
+      ctx.hp = EVP_aes_256_ctr();
+      return 0;
+    case 0x03001303u: // TLS_CHACHA20_POLY1305_SHA256
+      ctx.aead = EVP_chacha20_poly1305();
+      ctx.hp = EVP_chacha20();
+      return 0;
+    default:
+      return -1;
+  }
+}
+
+inline size_t AEAD_Max_Overhead(const CryptoContext& ctx) {
+  return aead_tag_length(ctx);
+}
+
+inline void ClearTLSError() {
+  ERR_clear_error();
+}
+
+inline const char* TLSErrorString(int code) {
+  return ERR_error_string(code, nullptr);
+}
+
+}  // namespace
+
+
+// Reset the QuicSessionConfig to initial defaults.
+// TODO(@jasnell): Currently only called once so candidate for inlining.
+void QuicSessionConfig::ResetToDefaults() {
+#define V(idx, name, def) name##_ = def;
+  QUICSESSION_CONFIG(V)
+#undef V
+}
+
+// Sets the QuicSessionConfig using an AliasedBuffer for efficiency.
+void QuicSessionConfig::Set(Environment* env) {
+  ResetToDefaults();
+  AliasedBuffer<double, Float64Array>& buffer =
+      env->quic_state()->quicsocketconfig_buffer;
+  uint64_t flags = buffer[IDX_QUIC_SOCKET_CONFIG_COUNT];
```

Comment:
Should this be `buffer[IDX_QUIC_SOCKET_CONFIG_COUNT - 1];` perhaps?

Reply:
No, there's an extra slot here to communicate which items are set and which are not, similar to how settings work in the http2 impl.